core/src/execution/client/rpc.rs

```diff
@@ -0,0 +1,609 @@
+            });
+
+            let account_chunk = join_all(account_chunk_futs).await;
+
+            for (address, value) in account_chunk {
+                let account = value?;
+                account_map.insert(address, account);
+            }
+        }
+
+        Ok(account_map)
+    }
+
+    async fn chain_id(&self) -> Result<u64> {
+        self.rpc.chain_id().await
+    }
+
+    async fn get_block(
+        &self,
+        block_id: BlockId,
+        full_tx: bool,
+    ) -> Result<Option<N::BlockResponse>> {
+        self.rpc.get_block(block_id, full_tx.into()).await
+    }
+
+    async fn get_block_receipts(
+        &self,
+        block_id: BlockId,
+    ) -> Result<Option<Vec<N::ReceiptResponse>>> {
+        self.rpc.get_block_receipts(block_id).await
+    }
+
+    async fn send_raw_transaction(&self, bytes: &[u8]) -> Result<B256> {
+        self.rpc.send_raw_transaction(bytes).await
+    }
+
+    async fn new_filter(&self, filter: &Filter) -> Result<U256> {
+        self.rpc.new_filter(filter).await
+    }
+
+    async fn new_block_filter(&self) -> Result<U256> {
+        self.rpc.new_block_filter().await
+    }
+
+    async fn new_pending_transaction_filter(&self) -> Result<U256> {
+        self.rpc.new_pending_transaction_filter().await
+    }
+
+    async fn uninstall_filter(&self, filter_id: U256) -> Result<bool> {
+        self.rpc.uninstall_filter(filter_id).await
+    }
+}
+
+impl<N: NetworkSpec, R: ExecutionRpc<N>> ExecutionInnerRpcClient<N, R> {
+    async fn verify_proof_to_account(
+        &self,
+        proof: EIP1186AccountProofResponse,
+        block: &N::BlockResponse,
+        verify_code: bool,
+    ) -> Result<Account> {
+        // Verify the account proof
+        verify_account_proof(&proof, block.header().state_root())?;
+        // Verify the storage proofs
+        verify_storage_proof(&proof)?;
+        // Verify the code hash
+        let code = if verify_code {
+            let code = self
+                .rpc
+                .get_code(proof.address, block.header().number().into())
+                .await?
+                .into();
+            verify_code_hash_proof(&proof, &code)?;
+            Some(code)
+        } else {
+            None
+        };
+
+        Ok(Account {
+            account: TrieAccount {
+                nonce: proof.nonce,
+                balance: proof.balance,
+                storage_root: proof.storage_hash,
+                code_hash: proof.code_hash,
+            },
+            code,
+            account_proof: proof.account_proof,
+            storage_proof: proof.storage_proof,
+        })
+    }
+
+    /// Verify the integrity of each log entry in the given array of logs by
+    /// checking its inclusion in the corresponding transaction receipt
+    /// and verifying the transaction receipt itself against the block's receipt root.
+    async fn verify_logs(&self, logs: &[Log]) -> Result<()> {
```

Comment:
A lot of the work here feels similar to what we would need to do in the verifiable api to. Can we get some better code reuse?

Reply:
Some part of the work is indeed duplicated, yes, but code reuse is not trivial here because in API service case we need `ExecutionRpc.get_block_receipts` whereas in RPC client case we need `ExecutionInnerRpcClient.get_block_receipts` (for verified response). And it's only a small portion of the logic i.e. fetching block receipts for a list of block numbers -- which is the same, not the entire fn.